File: typescript/cote/GJM/철인_2종_경기.ts
const TEST_CASES = 10
const N = 10
const MAX_RUN_SPEED = 13
const MAX_CYCLE_SPEED = 20
const TRACK_LENGTH = 100

type Speed = number
type TrackLength = number
type Record = number

class Player {
	accessor runSpeed: Speed;
	accessor cycleSpeed: Speed;

	constructor(runSpeed: Speed, cycleSpeed: Speed) {
		this.runSpeed = runSpeed;
		this.cycleSpeed = cycleSpeed;
	}

	runTime(runTrackLength: TrackLength): Record {
		return runTrackLength / this.runSpeed;
	} 

	cycleTime(cycleTrackLength: TrackLength): Record {
		return cycleTrackLength / this.cycleSpeed
	}

	time(runTrackLength: TrackLength, cycleTrackLength: TrackLength) {
		return this.runTime(runTrackLength) + this.cycleTime(cycleTrackLength)
	}
}

function initialize(): [Speed[], Speed[]] {
	const runSpeeds = new Array(N).fill(0)
		.map(_ => Math.random() * (MAX_RUN_SPEED * 2 / 5) + MAX_RUN_SPEED * 3 / 5 );
	const cycleSpeeds = new Array(N).fill(0)
		.map(_ => Math.random() * (MAX_CYCLE_SPEED * 2 / 5) + MAX_CYCLE_SPEED * 3 / 5);

	return [runSpeeds, cycleSpeeds];
}

function solve(runSpeeds: Speed[], cycleSpeeds: Speed[]): TrackLength {
	const players: Player[] = new Array(N - 1).fill(0)
		.map((_, i) => new Player(runSpeeds[i], cycleSpeeds[i]));

	const cheater: Player = new Player(runSpeeds[N - 1], cycleSpeeds[N - 1]);

	function diff(runTrackLength: TrackLength) : Record {
		const cycleTrackLength = TRACK_LENGTH - runTrackLength;

		const cheaterRecord = cheater.time(runTrackLength, cycleTrackLength)
		let bestRecord = Number.MAX_VALUE

		for (const player of players) {
			bestRecord = Math.min(bestRecord, player.time(runTrackLength, cycleTrackLength))
		}

		return bestRecord - cheaterRecord
	}

	let [lo, hi] = [0, TRACK_LENGTH]

	for (let i = 0; i < 100; i++) {
		const aab = (lo * 2 + hi) / 3;
		const abb = (lo + 2 * hi) / 3;

		if (diff(aab) < diff(abb)) {
			lo = aab
		} else {
			hi = abb
		}
	}
	
	return (lo + hi) / 2;
}

function main() {
	for (let i = 0; i < TEST_CASES; i++) {
		const [runSpeeds, cycleSpeeds] = initialize();
		console.log(solve(runSpeeds, cycleSpeeds).toFixed(3));
	}
}

main()

File: c++/cote/GJM/철인_2종_경기.cpp
#include <iostream> 
#include <random>
#include <vector>
#define TEST_CASES 10
#define N 10
#define SPEED 30
#define TRACK_LENGTH 100

using namespace std;

vector<double> initialize(default_random_engine& gen){
	
	normal_distribution<double> dis(SPEED, SPEED/5);
	vector<double> ret;

	for(int i=0; i<N; i++){
		ret.push_back(abs(dis(gen)));
	}

	return ret;
}

double time(double run, double run_speed, double cycle_speed){
	double cycle = TRACK_LENGTH -run;
	return run / run_speed + cycle / cycle_speed;
}

double diff(double r, vector<double>& run_speed, vector<double>& cycle_speed){
	int n = run_speed.size();
	double cheater = time( r, run_speed[n-1], cycle_speed[n-1]);
	double others = time( r, run_speed[0], cycle_speed[0]);

	for(int i=1; i< n-1; i++)
		others = min(others, time( r, run_speed[i], cycle_speed[i]));

	return others- cheater;
}

double solve(vector<double>& run_speed, vector<double>& cycle_speed){
	double lo = 0, hi = TRACK_LENGTH ;

	for(int i=0 ;i < 100; i++){
		double aab = (2*lo + hi)/3;
		double abb = (lo + 2*hi)/3;

		if(diff(aab,run_speed, cycle_speed) > diff(abb,run_speed,cycle_speed))
			hi = abb;
		else 
			lo = aab;
	}

	return (lo+hi)/2;
}

int main(void){
	
	default_random_engine gen(43);	
	
	for(int i=0; i< TEST_CASES; i++){
		vector<double> run_speed = initialize(gen);
		vector<double> cycle_speed = initialize(gen);

		cout << solve(run_speed, cycle_speed) << endl;

	}

	return 0;
}